src/main/java/hudson/tasks/junit/History.java

```diff
@@ -64,49 +75,440 @@ public TestObject getTestObject() {
+        LinearRegression lr = new LinearRegression(lrX, lrY);
+        ObjectNode lrSeries = mapper.createObjectNode();
+        series.add(lrSeries);
+        lrSeries.put("name", title);
+        lrSeries.put("type", "line");
+        lrSeries.put("symbol", "circle");
+        lrSeries.put("symbolSize", 0);
+        lrSeries.put("xAxisIndex", xAxisIndex);
+        lrSeries.put("yAxisIndex", yAxisIndex);
+        ArrayNode lrData = mapper.createArrayNode();
+        lrSeries.set("data", lrData);
+        ObjectNode lrStyle = mapper.createObjectNode();
+        lrSeries.set("itemStyle", lrStyle);
+        lrStyle.put("color", color);
+        ObjectNode lrAreaStyle = mapper.createObjectNode();
+        lrSeries.set("areaStyle", lrAreaStyle);
+        lrAreaStyle.put("color", "rgba(0, 0, 0, 0)");
+
+        for (int index = 0; index < history.size(); ++index) {
+            //lrData.add((float)Math.min(maxV, Math.max(minV, lr.predict(index))));
+            lrData.add((float)lr.predict(index));
+        }
+    }
+
+    private void createSplineTrend(ObjectMapper mapper, ArrayNode series, List<HistoryTestResultSummary> history, double[] lrX, double[] lrY, String title, String color, double minV, double maxV, int xAxisIndex, int yAxisIndex) {
+        if (history.size() < 200) {
+            return;
+        }
+        double windowSize = 100.0;
+        double rho = Math.min(1.0, 1.0 / Math.max(1, (history.size() / 2)));
+        if (rho > 0.75) {
+            return; // Too close to linear
+        }
+        SmoothingCubicSpline scs = new SmoothingCubicSpline(lrX, lrY, 0.001);
+        ObjectNode lrSeries = mapper.createObjectNode();
+        series.add(lrSeries);
+        lrSeries.put("name", title);
+        lrSeries.put("type", "line");
+        lrSeries.put("symbol", "circle");
+        lrSeries.put("symbolSize", 0);
+        lrSeries.put("xAxisIndex", xAxisIndex);
+        lrSeries.put("yAxisIndex", yAxisIndex);
+        ArrayNode lrData = mapper.createArrayNode();
+        lrSeries.set("data", lrData);
+        ObjectNode lrStyle = mapper.createObjectNode();
+        lrSeries.set("itemStyle", lrStyle);
+        lrStyle.put("color", color);
+        ObjectNode lrAreaStyle = mapper.createObjectNode();
+        lrSeries.set("areaStyle", lrAreaStyle);
+        lrAreaStyle.put("color", "rgba(0, 0, 0, 0)");
+
+        for (int index = 0; index < history.size(); ++index) {
+            //lrData.add((float)Math.min(maxV, Math.max(minV, scs.evaluate(index))));
+            lrData.add((float)scs.evaluate(index));
+        }
+    }
+
+    private ObjectNode computeResultTrendJson(List<HistoryTestResultSummary> history) {
+        ObjectMapper mapper = new ObjectMapper();
+        ObjectNode root = mapper.createObjectNode();
+        ArrayNode domainAxisLabels = mapper.createArrayNode();
+        ArrayNode series = mapper.createArrayNode();
+
+        ObjectNode okSeries = mapper.createObjectNode();
```

Comment:
would it be possible to extract some of this repetitive code to methods please? They all look the same except different name, color and data.

Colors should some from the design library as well for theming: 
https://weekly.ci.jenkins.io/design-library/Colors/

Reply:
Maybe, but this is also the very first iteration (I haven't seen any plugins use ECharts like this so far, I am not sure, are there examples). I don't know where this will go from here, so I would prefer not to over-streamline.

It's not so easy regarding theming for now, because I think every modern webpage should offer dark mode and last time I checked, Jenkins breaks it.

Reply:
Jenkins official color palette that I linked should work fine in dark mode

Reply:
How do I fetch color palette on Java side? Is that even possible? I guess I need to override the colors on Javascript side. If this is required to merge, I can do it next week.